d3a_interface/data_classes.py

```diff
@@ -0,0 +1,442 @@
+        self.id = str(id)
+        self.time = time
+        self.original_price = original_price or price
+        self.price = price
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
+        self.attributes = attributes
+        self.requirements = requirements
+
+    def update_price(self, price: float) -> None:
+        self.price = price
+        self.energy_rate = self.price / self.energy
+
+    def update_energy(self, energy: float) -> None:
+        self.energy = energy
+        self.energy_rate = self.price / self.energy
+
+    def to_json_string(self, **kwargs) -> str:
+        """Convert the Offer or Bid object into its JSON representation.
+
+        Args:
+            **kwargs: additional key-value pairs to be added to the JSON representation.
+        """
+        obj_dict = deepcopy(self.__dict__)
+        if kwargs:
+            obj_dict = {**obj_dict, **kwargs}
+
+        obj_dict["type"] = self.__class__.__name__
+
+        return json.dumps(obj_dict, default=my_converter)
+
+    def serializable_dict(self) -> Dict:
+        return {
+            "type": self.__class__.__name__,
+            "id": self.id,
+            "energy": self.energy,
+            "energy_rate": self.energy_rate,
+            "original_price": self.original_price,
+            "time": datetime_to_string_incl_seconds(self.time),
+            "attributes": self.attributes,
+            "requirements": self.requirements
+        }
+
+    @classmethod
+    def from_json(cls, offer_or_bid, current_time=None) -> Union["Offer", "Bid"]:
+        offer_bid_dict = json.loads(offer_or_bid)
+        object_type = offer_bid_dict.pop("type")
+
+        offer_bid_dict.pop("energy_rate", None)
+
+        if object_type == "Offer":
+            offer_bid_dict["time"] = current_time
+            return Offer(**offer_bid_dict)
+        if object_type == "Bid":
+            return Bid(**offer_bid_dict)
+
+
+class Offer(BaseBidOffer):
+    def __init__(self, id: str, time: datetime, price: float,
+                 energy: float, seller: str, original_price: Optional[float] = None,
+                 seller_origin: str = None, seller_origin_id: str = None,
+                 seller_id: str = None,
+                 attributes: Dict = None,
+                 requirements: List[Dict] = None):
+        super().__init__(id=id, time=time, price=price, energy=energy,
+                         original_price=original_price,
+                         attributes=attributes, requirements=requirements)
+        self.seller = seller
+        self.seller_origin = seller_origin
+        self.seller_origin_id = seller_origin_id
+        self.seller_id = seller_id
+
+    def __hash__(self) -> int:
+        return hash(self.id)
+
+    def __repr__(self) -> str:
+        return ("<Offer('{s.id!s:.6s}', '{s.energy} kWh@{s.price}', '{s.seller} {rate}'>"
+                .format(s=self, rate=self.energy_rate))
```

Comment:
Although its more like a migration from d3a to d3a-interface BUT still how about using f-string now

Reply:
@faizan2590 seems great, done